### ISSUE TITLE
devicemanager stuck in Allocate RPC and causing all new coming pods pending on admitting

### ISSUE THREAD

**Issue by ytinirt** [kubernetes/kubernetes#129249]
### What happened?

After created pod with device request, its state always be `Pending`, even though it was scheduled.
And at the same time, new coming pod (without device request) on the same node is on `Pending` state too.

After a bit investigation with goroutines, devicemanager was stucking in `Allocate` RPC, and wait device plugin's response:
![image](https://github.com/user-attachments/assets/e6af5d68-e668-45e1-b61d-5a8be3c01686)

Accroding to this stack, pod admitting is stucked obviously.

### What did you expect to happen?

Device plugins' bug or out-of-service, should not lead kubelet pod admit procedure blocking.

### How can we reproduce it (as minimally and precisely as possible)?

Reproduct it with https://github.com/kube-HPC/k8s-dummy-device-plugin. Add some sleep time in `Allocate` call:
![image](https://github.com/user-attachments/assets/9a1b021a-27a1-43a1-af73-0f6426e7f169)

![image](https://github.com/user-attachments/assets/501dd05e-10f8-4327-825b-16554896c0b6)

pod with dummy device and new coming pod without dummy device are in `Pending` state:
![image](https://github.com/user-attachments/assets/467d25c3-ca22-405d-8ece-c548b4deca52)

After running out sleep time, all pod return to `Running` state:
![image](https://github.com/user-attachments/assets/212bcc69-df04-490f-8fad-24c8be5e4ce2)

### Anything else we need to know?

NA

### Kubernetes version

<details>

```console
v1.25
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
CentOS
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>
cri-o
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by ffromani** [comment]
/triage accepted
/priority important-longterm

pretty clear and longstanding issue

**Comment by SuQiucheng** [comment]
I think I can add a timeout, when kubelet invokes the allocate interface of the plug-in

**Comment by bart0sh** [comment]
> And at the same time, new coming pod (without device request) on the same node is on Pending state too.

This is alarming. It means that all new pods scheduled to the node are stuck because of not responding device plugin? This looks like a serious issue and introducing rpc timeout is not the best way to fix it, I guess.

@ffromani @swatisehgal WDYT?

**Comment by ffromani** [comment]
> > And at the same time, new coming pod (without device request) on the same node is on Pending state too.
> 
> This is alarming. It means that all new pods scheduled to the node are stuck because of not responding device plugin, right? This looks like a serious issue and introducing rpc timeout is not the best way to fix it, I guess.
> 
> [@ffromani](https://github.com/ffromani) [@swatisehgal](https://github.com/swatisehgal) WDYT?

yep, this seems right and ties on the effort @SergeyKanzhelev is leading to improve the device reliability.
If memory serves, the only way to handle this is to detect containers stuck on the node and delete them explicitely using an extra controller. The device unavailability will pop up eventually on the kubelet and then to the scheduler, but the containers/pods stuck in this limbo will need to be deleted explicitely.

**Comment by swatisehgal** [comment]
A rogue device plugin or one with error impacting all pods is definitely a concerning issue. 

Can you help in providing more clarity on the scope of the impact: 

1. Does the system reconcile only after allocate() call succeeds after the sleep times out? 
2. Do all the pods go to the same node? Do you have a single node in the cluster?
3. It appears that you have standalone pods (not deployments) deployed, why do we see multiple pods with the same name in `ContainerCreating` state? What happens to them eventually?

I don't understand why this impacts other pods (even the ones that don't require devices). Assuming the advertisement of devices is working fine, pod should be scheduled to a node and during container creation stage Kubelet calls device plugin pod for device allocation. This is where the sleep in Allocate (which is indicative of some allocation issue) comes into play.

Pods being in pending means that the pod was not successfully scheduled to the node which seems to me like the devices were not advertised properly or there is resource contention issue.

Pod being pending is sign of pod not getting scheduled but the unsuccessful allocation in device plugin (simulated by sleep) happens at container creation stage. 

I am not able to connect the dots here, need to dig deeper into this.

**Comment by SuQiucheng** [comment]
After a pod is scheduled to a node, kubelet executes `canAdmitPod` to allocate devices. This process is serial for kubelet. You can imagine that if node resources are allocated concurrently for all pods, unexpected problems may occur. Therefore, when a pod is stuck in the canadmitpod, other pods cannot be processed by kubelet.

**Comment by bart0sh** [comment]
> Therefore, when a pod is stuck in the canadmitpod, other pods cannot be processed by kubelet.

It looks like the synchronous admission loop was designed before device plugins were designed as this approach doesn't expect significant admission delays. Time to re-design/reimplement?

**Comment by bart0sh** [comment]
@ytinirt @SergeyKanzhelev I believe it's closed by mistake. Please, reopen.
